### PR TITLE
Add `FirestoreGradesStateRepository`

### PR DESCRIPTION
**Pull request by Jonas-Sander**

### CODE REVIEW

File: app/lib/grades/grades_service/grades_service.dart [hunk 1]
```diff
@@ -680,7 +689,15 @@ class Grade {
...
 
-enum WeightType { perGrade, perGradeType, inheritFromTerm }
+enum WeightType {
+  perGrade,
+  perGradeType,
+  inheritFromTerm;
+
+  static WeightType fromString(String value) {
+    return values.firstWhere((e) => e.name == value);
+  }
```
Comment (nilsreichardt) [inline app/lib/grades/grades_service/grades_service.dart:699]: Why not `WeightType.values.byName(value)` in the code where you call `WeightType.fromString(value)`?

Comment (Jonas-Sander) [reply app/lib/grades/grades_service/grades_service.dart:699]: I forgot that this method exists, thanks for the reminder 👍

File: app/test/grades/grades_repository_test.dart [hunk 2]
```diff
@@ -0,0 +1,615 @@
...
+          weight: const Weight.percent(250));
+
+      final res = repository.data;
+      final state = repository.state.value;
+
+      expect(res, {
+        'currentTerm': '02-10-term',
+        'terms': {
+          '02-10-term': {
+            'id': '02-10-term',
```
Comment (nilsreichardt) [inline app/test/grades/grades_repository_test.dart:185]: Should we also store a `createdAt` for the term?

File: app/test/grades/grades_repository_test.dart [hunk 3]
```diff
@@ -0,0 +1,615 @@
...
+              'mathe': {
+                'value': 2.5,
+                'type': 'factor',
+              }
+            },
+            'gradeTypeWeights': {
+              'vocabulary-test': {'value': 1.5, 'type': 'factor'}
+            },
+            'subjects': {
+              'mathe': {
```
Comment (nilsreichardt) [inline app/test/grades/grades_repository_test.dart:198]: Same here, should add a `createdAt`?

File: app/test/grades/grades_repository_test.dart [hunk 4]
```diff
@@ -0,0 +1,615 @@
...
+                'weightType': 'inheritFromTerm',
+                'finalGradeType': 'oral-participation'
+              },
+            },
+            'finalGradeType': 'my-custom-grade-type'
+          }
+        },
+        'grades': {
+          'grade-1': {
+            'id': 'grade-1',
```
Comment (nilsreichardt) [inline app/test/grades/grades_repository_test.dart:258]: Same here, should add a `createdAt` for grades?

**Review by nilsreichardt — APPROVED** [review]
LGTM

**Pull request merged**